include/gridtools/fn/backend/gpu.hpp

```diff
@@ -23,108 +24,154 @@
 
 namespace gridtools::fn::backend {
     namespace gpu_impl_ {
+        template <class KeyValuePair, class = void>
+        struct is_valid_block_size_key_value_pair : std::false_type {};
+
+        template <template <class...> class List, class Key, class Value>
+        struct is_valid_block_size_key_value_pair<List<Key, Value>,
+            std::enable_if_t<is_integral_constant<Value>::value && (Value::value > 0)>> : std::true_type {};
+
+        template <class BlockSizes>
+        using is_valid_block_sizes =
+            std::bool_constant<meta::is_list<BlockSizes>::value &&
+                               meta::all<meta::transform<is_valid_block_size_key_value_pair, BlockSizes>>::value>;
+
         /*
-         * BlockSizes must be a meta map, mapping dimensions to integral constant block sizes.
+         * ThreadBlockSizes and LoopBlockSizes must be meta maps, mapping dimensions to integral constant block sizes.
          *
          * For example, meta::list<meta::list<dim::i, integral_constant<int, 32>>,
          *                         meta::list<dim::j, integral_constant<int, 8>>,
          *                         meta::list<dim::k, integral_constant<int, 1>>>;
          * When using a cartesian grid.
          */
-        template <class BlockSizes>
+        template <class ThreadBlockSizes, class LoopBlockSizes = meta::list<>>
         struct gpu {
-            using block_sizes_t = BlockSizes;
+            using thread_block_sizes_t = ThreadBlockSizes;
+            using loop_block_sizes_t = LoopBlockSizes;
             cudaStream_t stream = 0;
+
+            static_assert(is_valid_block_sizes<ThreadBlockSizes>::value, "invalid thread block sizes");
+            static_assert(is_valid_block_sizes<LoopBlockSizes>::value, "invalid loop block sizes");
         };
 
-        template <class BlockSizes, class Dims, int I>
-        using block_size_at_dim = meta::second<meta::mp_find<BlockSizes, meta::at_c<Dims, I>>>;
+        template <class BlockSizes>
+        struct block_size_at_dim {
+            template <class Dim>
+            using apply = std::conditional_t<std::is_void_v<meta::mp_find<BlockSizes, Dim, void>>,
+                meta::list<Dim, integral_constant<int, 1>>,
+                meta::mp_find<BlockSizes, Dim, void>>;
```

Comment:
```suggestion
            using apply = meta::mp_find<BlockSizes, Dim, meta::list<Dim, integral_constant<int, 1>>>;
```
?

Reply:
That’s what I tried first, but does not compile. However, I don’t remember anymore why. Will check.

Reply:
Looks like I was actually just too dumb to write the code correctly, does compile flawlessly.